src/backend/distributed/utils/causal_clock.c

```diff
@@ -0,0 +1,673 @@
+								connection->connectionId, connection->port,
+								nodeClockValue->logical, nodeClockValue->counter)));
+
+		if (globalClockValue)
+		{
+			globalClockValue = LargerClock(globalClockValue, nodeClockValue);
+		}
+		else
+		{
+			globalClockValue = nodeClockValue;
+		}
+
+		PQclear(result);
+		ForgetResults(connection);
+	}
+
+	ereport(DEBUG1, (errmsg("final global transaction clock %lu:%u",
+							globalClockValue->logical,
+							globalClockValue->counter)));
+	return globalClockValue;
+}
+
+
+/*
+ * AdjustClocksToTransactionHighest Sets the clock value of all the nodes, participated
+ * in the PREPARE of the transaction, to the highest clock value of all the nodes.
+ */
+static void
+AdjustClocksToTransactionHighest(List *nodeConnectionList, ClusterClock
+								 *transactionClockValue)
+{
+	StringInfo queryToSend = makeStringInfo();
+
+	/* Set the clock value on participating worker nodes */
+	appendStringInfo(queryToSend,
+					 "SELECT pg_catalog.citus_internal_adjust_local_clock_to_remote"
+					 "('(%lu, %u)'::pg_catalog.cluster_clock);",
+					 transactionClockValue->logical, transactionClockValue->counter);
+
+	ExecuteRemoteCommandInConnectionList(nodeConnectionList, queryToSend->data);
+}
+
+
+/*
+ * PrepareAndSetTransactionClock polls all the transaction-nodes for their respective clocks,
+ * picks the highest clock and returns it via UDF citus_get_transaction_clock. All the nodes
+ * will now move to this newly negotiated clock.
+ */
+static ClusterClock *
+PrepareAndSetTransactionClock(void)
+{
+	if (!EnableClusterClock)
+	{
+		/* citus.enable_cluster_clock is false */
+		ereport(NOTICE, (errmsg("GUC enable_cluster_clock is off")));
+		return NULL;
+	}
+
+	if (!IsTransactionBlock())
+	{
+		ereport(NOTICE,
+				(errmsg("must be called inside a transaction and before a commit")));
+		return NULL;
+	}
+
+	dlist_iter iter;
+	List *transactionNodeList = NIL;
+	List *nodeList = NIL;
+
+	/* Prepare the connection list */
+	dlist_foreach(iter, &InProgressTransactions)
+	{
+		MultiConnection *connection = dlist_container(MultiConnection, transactionNode,
+													  iter.cur);
+		WorkerNode *workerNode = FindWorkerNode(connection->hostname, connection->port);
+
+		/* Skip the node if we already in the list */
+		if (list_member_int(nodeList, workerNode->groupId))
+		{
+			continue;
+		}
+		else
+		{
+			nodeList = lappend_int(nodeList, workerNode->groupId);
+		}
+
+		RemoteTransaction *transaction = &connection->remoteTransaction;
+
+		Assert(transaction->transactionState != REMOTE_TRANS_NOT_STARTED);
+
+		/* Skip a transaction that failed */
+		if (transaction->transactionFailed)
+		{
+			continue;
+		}
+
+		/*
+		 * Check if any DML or DDL is executed over the connection on any
+		 * placement/table. If yes, add to the list, otherwise we skip
+		 * prepare since the connection didn't perform any write (read-only)
+		 */
+		if (ConnectionModifiedPlacement(connection))
```

Comment:
Good catch, I added this patch of code y'day to avoid duplication of getting clock from the same node, the list should be populated along with the poll list. Fixed it now.